hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/upgrade/OMUpgradeFinalizer.java

```diff
@@ -0,0 +1,328 @@
+    public Void call() throws OMException {
+      try {
+        emitStartingMsg();
+
+        for (OMLayoutFeature f : versionManager.unfinalizedFeatures()) {
+          finalizeFeature(f);
+          updateLayoutVersionInVersionFile(f);
+          versionManager.finalized(f);
+        }
+
+        emitFinishedMsg();
+      } finally {
+        isDone = true;
+      }
+      return null;
+    }
+
+    private void finalizeFeature(OMLayoutFeature feature)
+        throws OMException {
+      OmUpgradeAction action = feature.onFinalizeAction().orElse(NOOP);
+
+      if (action == NOOP) {
+        emitNOOPMsg(feature.name());
+        return;
+      }
+
+      putFinalizationMarkIntoVersionFile(feature);
+
+      emitStartingFinalizationActionMsg(feature.name());
+      try {
+        action.executeAction(ozoneManager);
+      } catch (Exception e) {
+        logFinalizationFailureAndThrow(e, feature.name());
+      }
+      emitFinishFinalizationActionMsg(feature.name());
+
+      removeFinalizationMarkFromVersionFile(feature);
+    }
+
+    private void updateLayoutVersionInVersionFile(OMLayoutFeature feature)
+        throws OMException {
+      int prevLayoutVersion = currentStoredLayoutVersion();
+
+      updateStorageLayoutVersion(feature.layoutVersion());
+      try {
+        persistStorage();
+      } catch (IOException e) {
+        updateStorageLayoutVersion(prevLayoutVersion);
+        logLayoutVersionUpdateFailureAndThrow(e);
+      }
+    }
+
+    private void putFinalizationMarkIntoVersionFile(OMLayoutFeature feature)
+        throws OMException {
+      try {
+        emitUpgradeToLayoutVersionPersistingMsg(feature.name());
+
+        setUpgradeToLayoutVersionInStorage(feature.layoutVersion());
+        persistStorage();
+
+        emitUpgradeToLayoutVersionPersistedMsg();
+      } catch (IOException e) {
+        logUpgradeToLayoutVersionPersistingFailureAndThrow(feature.name(), e);
+      }
+    }
+
+    private void removeFinalizationMarkFromVersionFile(OMLayoutFeature feature)
+        throws OMException {
+      try {
+        emitRemovingUpgradeToLayoutVersionMsg(feature.name());
+
+        unsetUpgradeToLayoutVersionInStorage();
+        persistStorage();
+
+        emitRemovedUpgradeToLayoutVersionMsg();
+      } catch (IOException e) {
+        logUpgradeToLayoutVersionRemovalFailureAndThrow(feature.name(), e);
+      }
+    }
+
+
+
+
+
+    private void setUpgradeToLayoutVersionInStorage(int version) {
+      ozoneManager.getOmStorage().setUpgradeToLayoutVersion(version);
+    }
+
+    private void unsetUpgradeToLayoutVersionInStorage() {
+      ozoneManager.getOmStorage().unsetUpgradeToLayoutVersion();
+    }
+
+    private int currentStoredLayoutVersion() {
+      return ozoneManager.getOmStorage().getLayoutVersion();
+    }
+
+    private void updateStorageLayoutVersion(int version) {
+      ozoneManager.getOmStorage().setLayoutVersion(version);
+    }
+
+    private void persistStorage() throws IOException {
```

Comment:
Can this be moved to the individual layout version update methods above, so that the caller does not need to invoke the persist explicitly every time? As a caller, if I do **storage**.updateStorageLayoutVersion(), I can expect it to be persisted, right? I don't have a strong opinion about this, just a suggestion.

Reply:
Storage and OMStorage chose to have an implementation for setters where the setter does not persist the change to the disk, these are the two that I have checked, and decided to remain consistent with this behaviour. This is the sole reason for controlling this from outside the Storage class implementation. Here it is extracted into a method to make it easier to follow the flow, and to hide the implementation detail of how to persist the storage from the reader when he reads the backbone of the internal logic of the methods with a higher layer in the abstractions.